runtime/extern.c

```diff
@@ -447,21 +521,28 @@ static void extern_rec(value v)
       }
       goto next_item;
     }
-    /* Check if already seen */
-    if (Color_hd(hd) == Caml_blue) {
-      uintnat d = obj_counter - (uintnat) Field(v, 0);
-      if (d < 0x100) {
-        writecode8(CODE_SHARED8, d);
-      } else if (d < 0x10000) {
-        writecode16(CODE_SHARED16, d);
+    /* Check if object already seen */
+    if (! (extern_flags & NO_SHARING)) {
+      h = Hash(v);
+      while (1) {
+        if (! bitvect_test(pos_table.present, h)) break;
+        if (pos_table.entries[h].obj == v) {
+          uintnat d = obj_counter - pos_table.entries[h].pos;
+          if (d < 0x100) {
+            writecode8(CODE_SHARED8, d);
+          } else if (d < 0x10000) {
+            writecode16(CODE_SHARED16, d);
 #ifdef ARCH_SIXTYFOUR
-      } else if (d >= (uintnat)1 << 32) {
-        writecode64(CODE_SHARED64, d);
+          } else if (d >= (uintnat)1 << 32) {
+            writecode64(CODE_SHARED64, d);
 #endif
-      } else {
-        writecode32(CODE_SHARED32, d);
+          } else {
+            writecode32(CODE_SHARED32, d);
+          }
+          goto next_item;
+        }
+        h = (h + 1) & pos_table.mask;
```

Comment:
I find it a bit strange to have internal details of the table implementation (linear probing here and the bitvect shotpath above) exposed in the user code. Is there a good reason not to turn this into a lookup function (possibly marked inline for performance)?

Reply:
The lookup function has a not-very-nice type when expressed in C, because morally it returns a sum type, Found of position-in-the-output | Notfound of hash-code-for-insertion.  I pushed a tentative implementation along these lines.  GCC inlines the function, generates slightly less good code than in the original PR, but it doesn't  make a significant performance difference.